v3_async_wip/v3_async_wip/mqtt_client.py

```diff
@@ -438,28 +615,191 @@ async def wait_for_disconnect():
-                    logger.debug("Waiting for disconnect response...")
-                    await disconnect_done
-                    logger.debug("Stopping Paho network loop")
-                    self._mqtt_client.loop_stop()
+                    logger.debug("Waiting for disconnect to complete...")
+                    async with self.disconnected_cond:
+                        await self.disconnected_cond.wait_for(lambda: not self.is_connected())
+                    logger.debug("Waiting for network loop to exit and clearing task")
+                    await self._network_loop
+                    self._network_loop = None
+                    # Wait slightly for tasks started by the on_disconnect handler to finish.
+                    # This will prevent warnings.
+                    # TODO: improve efficiency by being able to wait on something specific
+                    await asyncio.sleep(0.02)
                 elif rc == mqtt.MQTT_ERR_NO_CONN:
                     # This happens when we disconnect while already disconnected.
                     # In this implementation, it should only happen if Paho's inner state
                     # indicates we would like to be connected, but we actually aren't.
                     # We still want to do this disconnect however, because doing so changes
                     # Paho's state to indicate we no longer wish to be connected.
                     logger.debug("Early disconnect return (Already disconnected)")
-                    disconnect_done.cancel()
+                    if self._network_loop_running():
+                        # This block should never execute
+                        logger.warning("Network loop unexpectedly still running. Waiting")
+                        await self._network_loop
+                    logger.debug("Removing finished network loop task")
+                    self._network_loop = None
                 else:
                     # This block should never execute
-                    logger.error("Unexpected result from Paho disconnect. Doing nothing.")
-                    disconnect_done.cancel()
+                    logger.warning(
+                        "Unexpected rc {} from Paho .disconnect(). Doing nothing.".format(rc)
+                    )
 
             else:
                 logger.debug("Already disconnected!")
+
+    async def subscribe(self, topic):
+        """
+        Subscribe to a topic from the MQTT broker.
+
+        :param str topic: a single string specifying the subscription topic to subscribe to
+
+        :raises: ValueError if topic is None or has zero string length.
+        :raises: MQTTError if there is an error subscribing
+        """
+        try:
+            mid = None
+            logger.debug("Attempting subscribe")
+            # Using this lock allows us to be sure that the ACK won't come in before the
+            # Future can be added to the pending dictionary
```

Comment:
I love how you did this, but this comment is a little misleading. Maybe better to say "This lock postpones the code inside `complete_sub` from setting the subscribe result until after we can add the future to the mid tracker". Or something like that.  The important part is that this code works in concert with `complete_sub` and that's not particularly obvious.

Reply:
Yeah, you're right, I ought to improve the clarity